docs: Event streaming documentation (v13) (#626)
frappe/erpnext_com: erpnext_com/www/docs/user/manual/en/erpnext_integration/event_streaming.md

 <!-- add-breadcrumbs -->
# Event Streaming

> 

Reviewer: ```suggestion

> Introduced in Version 13
```

> Event Streaming enables inter site communications.

Reviewer: ```suggestion
Event Streaming enables inter site communications between two ore more sites.
```

From version 13, you can **subscribe** to Document Types and **stream** Documents between different sites.

> For Example: If you have 4-5 companies hosted on different sites, one among them is the main site where you want to do ledger posting and other sites, where the Sales Invoices are generated. You could use Event Streaming in this case. For this, your child company sites can subscribe to the main company site for Item, Customer and Supplier Document Types and the main company can subscribe to the child companies for Sales Invoices.

Reviewer: ```suggestion
For Example: Consider you have 4 companies hosted on different sites, one among them is the main site where you want to do ledger posting and on other sites, the Sales Invoices are generated. You can use Event Streaming in this case. For this, your child company sites can subscribe to the main company site for Item, Customer, and Supplier Document Types. The main Company in turn can subscribe to the child companies for Sales Invoices.
```

To access Event Streaming, go to:
> Home > Integrations > Event Streaming

## 1. Prerequisites

> Before creating an Event Producer, one same user needs to be created on both the sites which will be used to access the site and will act as an Event Subsciber. Make sure the user has the necessary permissions for creation, updation and deletion of the subscribed doctypes.

Reviewer: ```suggestion
Before creating an Event Producer, a common user needs to be created on both the sites which will be used to access the site and will act as an Event Subscriber. Make sure the user has the necessary permissions for creation, updation, and deletion of the subscribed DocTypes.
```

7. Save.
8. The API key and API secret are then set in the Event Producer document for Authentication.

    ![Event Producer](/docs/assets/img/setup/integrations/event-producer-doc.png)

### 2.2 Approve Event Consumer on the Event Producer site
1. After the Event Producer has been created, an Event Consumer automatically gets created on the other site. By default all the Subscribed Document Types have the status as 'Pending'. In order to enable the Event Consumer to consume the documents of these Document Types, their Status needs to be updated to 'Approved'.
2. Go to: **Home > Integrations > Event Streaming > Event Consumer**.
3. Once you open the Event Consumer document you will see all the Document Types that the consumer has subscribed to. Change the status from 'Pending' to 'Approved' for all the Document Types that you want to approve to be consumed. You can change the status to 'Rejected' if you do not want the documents of that Document Type to be consumed.
4. Save.

    ![Event Consumer](/docs/assets/img/setup/integrations/event-consumer-doc.png)

**Note**: Document updates for Subscribed Document Types won't be synced unless they are Approved.

> 

Reviewer: add ### 2.3 Offline access with single site
add details about having a site on localhost and one on the cloud, so users can have 'offline ERP'.

## 3. Features

### 3.1 Event Update Log

> Event Update Log logs every create, update, delete for documents that have consumers on the Event Producer site.

Reviewer: ```suggestion
Event Update Log logs every create, update, and delete for documents that have consumers on the Event Producer site.
```

In order to view the Event Update Log:

Go to: **Home > Integrations > Event Streaming > Event Update Log**.

> - For 'Create' type the Update Type, Document Type, Document Name and the entire document (as json) is logged.

Reviewer: ```suggestion
- For 'Create' type the Update Type, Document Type, Document Name and the entire document (as JSON) is logged.
```

- For 'Update' type the Update Type, Document Type, Document Name and the updated data (difference between the previous state and current state of the document) is logged.

> - For 'Delete' type only the Update Type, Document Type and Document Name is logged.

Reviewer: ```suggestion
- For 'Delete' type only the Update Type, Document Type, and Document Name is logged.
```

![Event Update Log](/docs/assets/img/setup/integrations/event-update-log-doc.png)

### 3.2 Event Sync Log
Like the Update Log, Event Sync Log logs every document synced from the Event Producer on the Event Consumer site.
In order to view the Event Sync Log:

Go to: **Home > Integrations > Event Streaming > Event Sync Log**.

![Event Sync Log](/docs/assets/img/setup/integrations/event-sync-log.png)

A successfully synced event generates a log document with:

- **Update Type**: Create, Update or Delete
- **Status**: Sync Status
- **DocType**

> - **Event Producer**: The site URL from where the document was synced

Reviewer: from where the document was created?

- **Use Same Name**
- **Data**: The document data as JSON

    ![Event Sync Log](/docs/assets/img/setup/integrations/event-synced.png)

A failed event generates a log doc with the above fields along with:

- **Error**: The error because of which the document was not synced.
    ![Event Synced](/docs/assets/img/setup/integrations/event-failed-error.png)

- **Resync Button**: It also provides a 'Resync' button in order to resync the failed event.
    ![Event Failed](/docs/assets/img/setup/integrations/event-failed.png)

### 3.3 Dependency Syncing
Certain Document Types have dependencies. For example, before syncing a Sales Invoice, the Item and Customer need to be present in the current site. So, Item and Customer are dependencies for Sales Invoice. Event Streaming handles this by on-demand dependency syncing. Whenever any document is to be synced, it first checks whether the document has any dependencies (Link fields, Dynamic Link fields, Child Table fields, etc.). If that dependency is not fullfilled i.e. the dependent document (eg: Item) is not present on your consumer site, it will be synced first and then the Sales Invoice will be synced.

> For example: Sales Invoice syncing with Item dependency

Reviewer: ```suggestion
For example: Sales Invoice syncing with Item dependency:
```

    ![Event Dependency](/docs/assets/img/setup/integrations/event-dependency-sync.gif)

### 3.4 Naming Configuration
Check the 'Use Same Name' checkbox to let the documents have same name on both Event Producer and Event Consumer sites. If this is not checked, then the document will be created using the naming conventions of the current site.

![Use Same Name Config](/docs/assets/img/setup/integrations/event-use-same-name.png)

> **Note**: For Document Types that have naming series, it is adviced to keep the 'Use Same Name' checkbox unchecked, to prevent naming conflicts. If this is unchecked, the Documents are created by following the naming conventions on the current site and the 'Remote Site Name' and 'Remote Document Name' custom fields are set in the synced document to store the Event Producer site URL and the document name on the remote site respectively.

Reviewer: ```suggestion
> **Note**: For Document Types that have naming series, it is advised to keep the 'Use Same Name' checkbox unchecked, to prevent naming conflicts. If this is unchecked, the Documents are created by following the naming conventions on the current site and the 'Remote Site Name' and 'Remote Document Name' custom fields are set in the synced document to store the Event Producer site URL and the document name on the remote site respectively.
```

![Subscribed Document](/docs/assets/img/setup/integrations/event-subscribed-doc.png)

### 3.5 Mapping Configuration
If you want to stream documents between an ERPNext instance and another Frappe app for a particular Document Type which has the same structure but fieldnames are different in both the sites, you can use Event Streaming with Mapping Configuration.
For this you need to first set up a Document Type Mapping:

- **Mapping Name**: Give a unique name to the mapping
- **Local Document Type**: The Document Type in your current site

> - **Remote Document Type**: The Document Type on the Event Producer site which you want to sync.

Reviewer: ```suggestion
- **Remote Document Type**: The Document Type on the Event Producer site which you want to sync
```

In the Field Mapping child table:

- **Local Fieldname**: The fieldname in the Local Document type of your current site.

> - **Remote Fieldname**: The fieldname in the Remote Document type of the Event Producer site which you want to map to the Local Fieldname. During the sync the value of the remote fieldname gets copied to the local fieldname.

Reviewer: ```suggestion
- **Remote Fieldname**: The fieldname in the Remote Document type of the Event Producer site which you want to map to the Local Fieldname. During the sync, the value of the remote fieldname gets copied to the local fieldname.
```
